src/mono/wasm/runtime/crypto-worker.ts

```diff
@@ -0,0 +1,211 @@
+    private get STATE_RESP_P(): number { return 4; } // Response has multiple parts
+    private get STATE_AWAIT(): number { return 5; } // Awaiting the next part
+
+    private constructor(msg_char_len: number) {
+        this.msg_char_len = msg_char_len;
+
+        const int_bytes = 4;
+        const comm_byte_len = int_bytes * (this.COMM_LAST_IDX + 1);
+        this.comm_buf = new SharedArrayBuffer(comm_byte_len);
+
+        // JavaScript character encoding is UTF-16.
+        const char_bytes = 2;
+        const msg_byte_len = char_bytes * this.msg_char_len;
+        this.msg_buf = new SharedArrayBuffer(msg_byte_len);
+
+        // Create the local arrays to use.
+        this.comm = new Int32Array(this.comm_buf);
+        this.msg = new Uint16Array(this.msg_buf);
+    }
+
+    public get_msg_len(): number { return this.msg_char_len; }
+    public get_msg_buffer(): SharedArrayBuffer { return this.msg_buf; }
+    public get_comm_buffer(): SharedArrayBuffer { return this.comm_buf; }
+
+    public send_msg(msg: string): string {
+        if (Atomics.load(this.comm, this.STATE_IDX) !== this.STATE_IDLE) {
+            throw "OWNER: Invalid sync communication channel state. " + Atomics.load(this.comm, this.STATE_IDX);
+        }
+        this.send_request(msg);
+        return this.read_response();
+    }
+
+    public shutdown(): void {
+        if (Atomics.load(this.comm, this.STATE_IDX) !== this.STATE_IDLE) {
+            throw "OWNER: Invalid sync communication channel state. " + Atomics.load(this.comm, this.STATE_IDX);
+        }
+
+        // Notify webworker
+        Atomics.store(this.comm, this.MSG_SIZE_IDX, 0);
+        Atomics.store(this.comm, this.STATE_IDX, this.STATE_SHUTDOWN);
+        Atomics.notify(this.comm, this.STATE_IDX);
+    }
+
+    private send_request(msg: string): void {
+        let state;
+        const msg_len = msg.length;
+        let msg_written = 0;
+
+        for (; ;) {
+            // Write the message and return how much was written.
+            const wrote = this.write_to_msg(msg, msg_written, msg_len);
+            msg_written += wrote;
+
+            // Indicate how much was written to the this.msg buffer.
+            Atomics.store(this.comm, this.MSG_SIZE_IDX, wrote);
+
+            // Indicate if this was the whole message or part of it.
+            state = msg_written === msg_len ? this.STATE_REQ : this.STATE_REQ_P;
+
+            // Notify webworker
+            Atomics.store(this.comm, this.STATE_IDX, state);
+            Atomics.notify(this.comm, this.STATE_IDX);
+
+            // The send message is complete.
+            if (state === this.STATE_REQ)
+                break;
+
+            // Wait for the worker to be ready for the next part.
+            //  - Atomics.wait() is not permissible on the main thread.
+            do {
+                state = Atomics.load(this.comm, this.STATE_IDX);
+            } while (state !== this.STATE_AWAIT);
```

Comment:
1. Should this, and the other similar spin-wait have a timeout too, in case the worker fails?
2. And should it also check for `STATE_SHUTDOWN`?

Reply:
I don't think so. The entire idea of this is the call is a synchronous operation and the caller is blocked waiting for a response. Hangs here are bugs that need to be fix and if we timeout or throw then we risk disabling the code path and falling back to the managed impl. I realize that may be worrisome, but the idea that we can timeout and then succeed is going to hide a bug and that seems like the wrong thing to do.